cmd/snap-repair/runner.go

```diff
@@ -67,24 +82,139 @@ func (r *Repair) SetStatus(status RepairStatus) {
 	}
+	defer statusR.Close()
+	defer statusW.Close()
+
+	// run the script
+	env := os.Environ()
+	// we need to hardcode FD=3 because this is the FD after
+	// exec.Command() forked. there is no way in go currently
+	// to run something right after fork() in the child to
+	// know the fd. However because go will close all fds
+	// except the ones in "cmd.ExtraFiles" we are safe to set "3"
+	env = append(env, "SNAP_REPAIR_STATUS_FD=3")
+	env = append(env, "SNAP_REPAIR_RUN_DIR="+rundir)
+
+	workdir := filepath.Join(rundir, "work")
+	if err := os.MkdirAll(workdir, 0700); err != nil {
+		return err
+	}
+
+	cmd := exec.Command(script)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	cmd.Env = env
+	cmd.Dir = workdir
+	cmd.ExtraFiles = []*os.File{statusW}
+	cmd.Stdout = logf
+	cmd.Stderr = logf
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+	statusW.Close()
 
-	// XXX actually run things and captures output etc
+	// wait for repair to finish or timeout
+	var scriptErr error
+	killTimerCh := time.After(defaultRepairTimeout)
+	doneCh := make(chan error)
+	go func() {
+		doneCh <- cmd.Wait()
+		close(doneCh)
+	}()
+	select {
+	case scriptErr = <-doneCh:
+		// done
+	case <-killTimerCh:
+		if err := osutil.KillProcessGroup(cmd); err != nil {
+			logger.Noticef("cannot kill timed out repair %s: %s", r, err)
+		}
+		scriptErr = fmt.Errorf("repair did not finish within %s", defaultRepairTimeout)
+	}
+	// read repair status pipe, use the last value
+	status := readStatus(statusR)
+	statusPath := filepath.Join(rundir, baseName+"."+status.String())
+
+	// if the script had an error exit status still honor what we
+	// read from the status-pipe, however report the error
+	if scriptErr != nil {
+		scriptErr = fmt.Errorf("%q failed: %s", r.Ref(), scriptErr)
+		if err := r.errtrackerReport(scriptErr, logPath); err != nil {
+			logger.Noticef("cannot report error to errtracker: %s", err)
+		}
+		// ensure the error is present in the output log
+		fmt.Fprintf(logf, "\n%s", scriptErr)
+	}
+	if err := os.Rename(logPath, statusPath); err != nil {
+		return err
+	}
+	r.SetStatus(status)
 
 	return nil
 }
 
+func readStatus(r io.Reader) RepairStatus {
+	var status RepairStatus
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		switch strings.TrimSpace(scanner.Text()) {
+		case "done":
+			status = DoneStatus
+		case "skip":
+			status = SkipStatus
+		}
+	}
+	if scanner.Err() != nil {
+		return RetryStatus
+	}
+	return status
+}
+
+// errtrackerReport reports an repairErr with the given logPath to the
+// snap error tracker.
+func (r *Repair) errtrackerReport(repairErr error, logPath string) error {
+	errMsg := fmt.Sprintf("%s", repairErr)
+
+	scriptOutput, err := ioutil.ReadFile(logPath)
+	if err != nil {
+		logger.Noticef("cannot read %s", logPath)
+	}
+	dupSig := fmt.Sprintf("%s\noutput:\n%s", errMsg, scriptOutput)
+
+	extra := map[string]string{
+		"Revision": strconv.Itoa(r.Revision()),
+		"BrandID":  r.BrandID(),
+		"RepairID": r.RepairID(),
```

Comment:
given that we can get here even with status Done now,  we want something like:

"Status": status.String()

as well